### DISCUSSION TITLE
[BUG]

### DISCUSSION THREAD

**Discussion by devwheel** [pnp/powershell#3009] [Q&A]
On windows 11 - upgraded to latest powershell - installed pnp.powershell 
Can't Import PnP.Powershell as I get the error:

PS C:\Users\admin> import-module PnP.PowerShell
import-module : Could not load file or assembly 'System.Management.Automation, Version=7.2.0.0, Culture=neutral,
PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
At line:1 char:1
+ import-module PnP.PowerShell
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Import-Module], FileNotFoundException
    + FullyQualifiedErrorId : System.IO.FileNotFoundException,Microsoft.PowerShell.Commands.ImportModuleCommand

**Comment by veronicageek** [comment]
That's because you've updated the module in PowerShell 5, which is not supported. 
If you want to use the latest `2.1.x`, then you'll need to be on PowerShell 7.2 or later.

**Comment by AndersRask** [comment]
@devwheel use this guide to move from older versions of pnp and PowerShell
https://github.com/pnp/powershell/blob/dev/MIGRATE-1.0-to-2.0.md

**Comment by davokr** [comment]
Quite frankly, none of these are solutions as Azure Automation's support for PowerShell 7.1 & 7.2 is still in preview.

**Reply by AndersRask** [reply]
7.3 is in preview, 7.1 is out of support and 7.2 is LTS

EDIT: ah for Automation accounts. Well yes, but 7.1 was in preview for years, so not sure how serious I can take that :)

**Reply by davokr** [reply]
> 7.3 is in preview, 7.1 is out of support and 7.2 is LTS
> 
> EDIT: ah for Automation accounts. Well yes, but 7.1 was in preview for years, so not sure how serious I can take that :)

Quite frankly I blame Microsoft, if they had implemented a native way to limit modules based on PSVersion compatibility, issues like this could have been avoided.

7.1 & 7.2 are still in preview in Azure Automation, 7.3 is unavailable last I checked, there's still no support for Source Control to handle runtime versions, or even changing the runtime version of an existing runbook except for using an API call.

**Reply by AndersRask** [reply]
PowerShell 7.2 and Python 3.10 was added as preview in october 2022 yes.

Could you elaborate what you mean with "limit modules based on PSVersion compatibility"?
Are you talking in Automation Accounts or in PowerShell in general? 
I think both have support for specifying PowerShell version?

Not sure what you mean with the source control :)

**Reply by davokr** [reply]
> PowerShell 7.2 and Python 3.10 was added as preview in october 2022 yes.
> 
> Could you elaborate what you mean with "limit modules based on PSVersion compatibility"? Are you talking in Automation Accounts or in PowerShell in general? I think both have support for specifying PowerShell version?
> 
> Not sure what you mean with the source control :)

Regarding PSVersion Compatibility, I think it would have been beneficial for Microsoft to have implemented defining the minimum/maximum version compatibility of a module version against $PSVersionTable.PSVersion. ie, PnP.PowerShell Major version = 3 requires $PSVersionTable.PSVersion Minimum Major Version = 7. If this could be properly defined per version of a module on the PowerShell Gallery, you could essentially _hide_ module versions that are incompatible with the version of PowerShell querying the gallery. This is for PowerShell in general.

Regarding Azure Automation, things are somewhat limited as Microsoft, while having support to publish PowerShell v7.1 & v7.2, does not (by default) publish new runbooks that are synchronized from the configure "Source Control" (github/devops) as anything but v5.1.

This applies to the general discuss around PnP.PowerShell & Azure Automation in the following way:

If you publish a new runbook via GitHub/DevOps integration to an Azure Automation account, the runbook is generated as v5.1 runtime. When you attempt to add the PnP.PowerShell module from the Module gallery, it does not provide minimum/maximum version control as you would see in Install-Module. Because of this, it attempts to install the latest version. However, when adding a PowerShell module to the Azure Automation account, you also have to specify the runtime version that the module will be installed in the context of. As your runbook is in v5.1, the module must also be installed in the context of v5.1. Due to the incompatibility of PnP.PowerShell's v2.* requiring PowerShell v7.2 or later, the module import will fail.

The context of this really applies more to #2993

**Reply by AndersRask** [reply]
You can specify version (and required modules, clr version etc) in the module manifest and in #requires statements in the module itself.

https://learn.microsoft.com/en-us/powershell/scripting/developer/module/how-to-write-a-powershell-module-manifest?view=powershell-7.3#module-manifest-elements
https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-7.3

PowerShell gallery is just a repository like any other. You can use Find-Module with version parameters if the limited web site UI does not suffice (never really use the web site I must admit).

I have never used the build in source control of AAC. We publish everything through CI/CD pipelines or PowerShell deployments, as we want 100% control with what is published, when and how.

I agree that it makes little sense that 7.1, that is not even a supported version anymore, is still in preview (and has been for 1½ year). MS should definitely push for getting 7.2 the golden standard.
